### ISSUE TITLE
Ubuntu/Debian guest on macOS High Sierra shows wrong files in NFS folders.

### ISSUE THREAD

**Issue by Yaroon** [hashicorp/vagrant#8957]
This is just an issue to direct attention again to https://github.com/mitchellh/vagrant/issues/8788 which got closed.

Does anyone on macOS High Sierra see this?

The Vagrantfile is simple enough with a www folder synced via NFS.

```
Vagrant.configure("2") do |config|
  config.vm.box = "hashicorp/precise64"
  config.vm.network "private_network", type: "dhcp"
  config.vm.synced_folder "www", "/var/www", type: "nfs"
end
```

Once vagrant is up, ssh into it and cd into the /var/www folder.

Run the following command:

`mkdir test && cd test && touch test{1..1000} && ls | wc -l`

It should show 1000, but in my case it shows 1038 which means there's something wrong...
When I run ls | uniq -c | grep " 2" it shows several files that appear twice. There's also files that do not show up at all.

**Comment by oxyc** [comment]
Does it work with NFS over TCP?

```rb
Vagrant.configure("2") do |config|
  config.vm.box = "hashicorp/precise64"
  config.vm.network "private_network", type: "dhcp"
  config.vm.synced_folder "www", "/var/www", type: "nfs", nfs_udp: false
end
```

**Comment by Yaroon** [comment]
That gives me the same result.

It is a problem indeed with Drupal 8. It has a lot of config files in 1 folder. As soon as you get 200 files or more it starts to miss some files in the folder and others are reported doubly.

This is the bug I believe that causes it. But I don't see it when I had macOS Sierra, now in High Sierra it is buggy.

https://superuser.com/questions/640419/how-can-i-have-two-files-with-the-same-name-in-a-directory-when-mounted-with-nfs

Also when I mount the folder over NFS from another Mac with macOS Sierra, it shows the right amount of files. So far I only see it fail when when mounting a guest OS on the same machine. Vagrant doesn't even need to be involved. Just booting the OS in VirtualBox has the same problem.

I don't know if it's a bug in macOS High Sierra (which does use APFS) or if it's a bug in Ubuntu and Debian. The superuser link above might suggest it's a bug in the Linux kernel... but that seems to have already been fixed.

In any case, if you have macOS High Sierra. I'd love it if you could just vagrant up that simple Vagrantfile and see for yourself what it reports when creating 1000 files in a folder.

**Comment by oxyc** [comment]
Unfortunately I don't–I just figured it was worth a try as UDP has been causing some weird issues before.

So to direct more attention #8788. Drupal VM has 2 reports from High Sierra users with this issue as well. https://github.com/geerlingguy/drupal-vm/issues/1547

**Comment by chrisroberts** [comment]
Does the same behavior persist if you set `nfs_version: 4` and `nfs_udp: false` ?

**Comment by oxyc** [comment]
Is NFSv4 available on macOS in High Sierra? At least it's not in Sierra. In https://github.com/geerlingguy/drupal-vm/issues/1547#issuecomment-330033200 @ajhoddinott verified that TCP does not fix it.

**Comment by chrisroberts** [comment]
I'm unsure. I'm going to be pulling high sierra today so I can generate a new box for testing, but it likely won't be ready until tomorrow.

**Comment by Yaroon** [comment]
```$ rpcinfo -p localhost | grep nfs
    100003    2   udp   2049  nfs
    100003    3   udp   2049  nfs
    100003    2   tcp   2049  nfs
    100003    3   tcp   2049  nfs```

macOS High Sierra does not offer NFS version 4, only 2 and 3 over both UDP and TCP.

**Comment by pascaldevink** [comment]
NFS version 4 is indeed not supported on High Sierra.

After being prompted. to upgrade to the latest version of vagrant and virtual box today, I decided to try this experiment again.
Using UDP it still fails in the exact same way. With TCP however, I get somewhat better results: 1007 files instead of 1000. This slightly improved sync seemed to be enough for my own project to be useful, although TCP is quite a bit slower than UDP.

I there some way we could try another installation of NFS, via homebrew for example? Maybe that could give us a more updated version, or support for version 4, just to try?

**Comment by oxyc** [comment]
I apologize for going offtopic

> although TCP is quite a bit slower than UDP.

I'd be interested in any numbers you have on this. What specific task was slower. For [myself](https://github.com/geerlingguy/drupal-vm/pull/1482#issuecomment-316821259) and others it turned out to be mostly equal and a bit faster even.

**Comment by Yaroon** [comment]
@pascaldevink I don't think it's acceptible to have 1007 files instead of 1000. It's not that it sees 7 files doubly, it's perfectly possible it sees 10 files twice and does not show 3 other files.

If you can't depends on the files being the exact same ones on an NFS volume it loses all purpose.

Today is macOS High Sierra release day. We'll see and hear a lot more about this very soon.

**Comment by mfn** [comment]
Hello,

> It should show 1000, but in my case it shows 1038 which means there's something wrong...
When I run ls | uniq -c | grep " 2" it shows several files that appear twice. There's also files that do not show up at all.

Can confirm, tested this in my guest, exact same result regarding the number of files 1038 appearing.

> We'll see and hear a lot more about this very soon.

I'll join the club.

> As soon as you get 200 files or more it starts to miss some files in the folder and others are reported doubly.

I'm seeing the same problems: in a particular directory I've 59 files in the host system and the guest shows me 60 files: one is completely missing and two appear twice `¯\_(ツ)_/¯`

In this particular VM I've **a lot** of code shared via NFS, the only viable solution at my place:
- files on host: 420520
- files visible in guest: 407604

Files seem to miss randomly. Elixir source files, node module js files,  git object hash files, files appearing twice. Here's a random diff screenshot (sorry but I think it's easier to see it instead of pasting text):

Left side: host, right side: guest ; the highlighted one on the right (guest) appears twice if you look at line `1283` too.
![image](https://user-images.githubusercontent.com/87493/30848584-e523f170-a29f-11e7-9e74-4d61712564bb.png)

### Environment
- The host is `maxOS High Sierra Version 10.13` on a `MacBook Pro (Retina, 15-inch, Late 2013)`
- The guest is `Linux swatio-dev 3.13.0-96-generic #143-Ubuntu SMP Mon Aug 29 20:15:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux` (Ubuntu 14.04 LTS)
- The config in `Vagrantfile` is: `config.vm.synced_folder ".", "/vagrant", type: "nfs"`
- I'm using my own [bento box](https://github.com/chef/bento), based on an older version of https://github.com/chef/bento/blob/master/ubuntu/ubuntu-14.04-amd64.json with a bigger root disk

### TL;DR
The workaround mentioned in https://github.com/geerlingguy/drupal-vm/issues/1547#issuecomment-330572626 worked for me, too

@Yaroon maybe you can link to this comment directly from your description?

**Comment by BrianGilbert** [comment]
This is a real bug.. I've reported it to Apple it's been there since early betas, here is a video that illustrates it..: https://www.dropbox.com/s/basc7us3frt7a7d/2017-09-25%20APFS%20NFS%20Issue.mp4

To fix the issue temporarily you can backup your machine with Carbon Copy Cloner and then format your drive back to HFS+ and restore the CCC backup to your drive

**Comment by thieman** [comment]
@BrianGilbert do you have a link to a radar or anything?

**Comment by chrisfromredfin** [comment]
I had an issue with High Sierra NFS not listing a file inside my VM (NFS is mounting from the Mac INTO an Ubuntu 14.04 guest - files reside on Mac, not inside VM shared out). I could cat the file, but ls wouldn't show it. I had mounted my NFS share with mount_options 'fsc' and 'noactime=2' -- removing those and running vagrant reload fixed it for me (or seemingly). Curious if anyone else is running with one of those caching options, and if disabling fixes.

**Comment by BrianGilbert** [comment]
@thieman not sure if you’ll be able to access it,but: https://bugreport.apple.com/web/?problemID=32866134

**Comment by Yaroon** [comment]
You can't access it, but I had filed it as https://bugreport.apple.com/web/?problemID=34127784 in August. It hasn't been marked as open or duplicate. I basically think it hasn't been looked at yet. Maybe because I categorised it as "Server"... could be on the backburner because of that.

**Comment by BrianGilbert** [comment]
@Yaroon I logged it 20th June and have had responses from Apple Engineers on the issue..

**Comment by jfbibeau** [comment]
Developer here upgraded to High Sierra before we could warn not to and APFS+NFS is completely screwed. Vagrant view of the mounted filesystem reports missing/duplicate files like this issue, and our gradle builds are completely non functional.

RHEL7 guest in our case, but same issue.

**Comment by BrianGilbert** [comment]
@jfbibeau you can use the trial of Carbon Copy Cloner to backup your machine to an external drive, then boot to Recovery mode and format the internal drive as HFS+.

Then boot from the external drive and use carbon copy cloner to restore back to your internal drive.. this will fix the issue.

**Comment by jfbibeau** [comment]
Thanks, I'll forward these instructions to him!

**Comment by navneet0693** [comment]
I did faced this problem when I tried macOS High Sierra beta. Here is a small write up about the same:

https://goo.gl/Xy6g3L

IRC: navneet0693
#drupal

**Comment by Yaroon** [comment]
macOS High Sierra 10.13.1 fixes this problem.
Currently the beta version is 10.13.1 Beta (17B25c)

**Comment by pascaldevink** [comment]
Unfortunately, macOS 10.13.1 beta 1 doesn't seem to fix it 😞

**Comment by chrisroberts** [comment]
Hi all. I'm closing this up simply because it's a duplicate. If you have any relevant information to add, please do so in #8788.

Cheers!
